ImageButton border (BorderWidth) overlaps the image instead of adding space - fix

kubaflo opened the pull request:
### Issues Fixed

Fixes https://github.com/dotnet/maui/issues/24856
Fixes https://github.com/dotnet/maui/issues/22520

|Before|After|
|--|--|
|<img src="https://github.com/user-attachments/assets/c3ba4ada-30a2-4c1a-a6bd-2570beba71f4" width="300px"/>|<img src="https://github.com/user-attachments/assets/ddb08607-410a-48e3-b3a2-976f40a5cbe1" width="300px"/>|

jsuarezruiz commented on src/Controls/tests/UITests/Tests/Issues/Issue20801.cs:
```diff
@@ -0,0 +1,25 @@
+		_ = App.WaitForElement("imageButton");
+
+		// The test passes if corners are properly rounded
+		VerifyScreenshot();
```
Could we run this tests in all the platforms?. The corners are usually the most fragile property, if we have more tests validating it on all platforms, better.

kubaflo replied:
Sure, done!

jsuarezruiz commented:
/azp run

jsuarezruiz commented:
/azp run MAUI-UITests-public

IrynaDoroshenkoDev commented:
Thank you Kuba! I used your fix in my handler and it works perfectly!

jsuarezruiz commented:
/azp run

jsuarezruiz commented:
/azp run

jsuarezruiz commented on src/Core/src/Platform/Android/ImageButtonExtensions.cs:
```diff
@@ -40,6 +40,9 @@ public static void UpdateCornerRadius(this ShapeableImageView platformButton, IB
 				.Build();
+
+			int contentPadding = (int)(radius / 2);
+			platformButton.SetContentPadding(contentPadding, contentPadding, contentPadding, contentPadding);
```
The changes are impacting the test Issue18242Test clipping the image:
![image](https://github.com/dotnet/maui/assets/6755973/48b33951-6273-42f3-91eb-fa24a95d20eb)

kubaflo replied:
Hmmm I'm not sure what is wrong here and if it is a good thing or not

jsuarezruiz commented:
/azp run

mattleibow commented:
@kubaflo I like these tests but I had to re-work the whole background/border/padding parts of the ImageButton because we lost the ripple at some point: https://github.com/dotnet/maui/pull/22298

I think I fixed the issue and my tests seemed to pass. But, not sure if you rebase but keep the tests if you get the same result.

kubaflo commented:
> @kubaflo I like these tests but I had to re-work the whole background/border/padding parts of the ImageButton because we lost the ripple at some point: #22298
> 
> I think I fixed the issue and my tests seemed to pass. But, not sure if you rebase but keep the tests if you get the same result.

@mattleibow I've merged the main branch. I've also added one commit because when I tested with the main branch the Android didn't work IMO. Please have a look:

|iOS|Android|
|--|--|
|<img src="https://github.com/dotnet/maui/assets/42434498/f6a7fdb6-2c88-4772-8696-4e3a8d98385c" width="300px"/>|<img src="https://github.com/dotnet/maui/assets/42434498/bda6e46b-3705-4e60-aff2-40aa35c07b77" width="300px"/>|

After my commit this is how the Android looks like:
|<img src="https://github.com/dotnet/maui/assets/42434498/3652e07d-44c5-4b5d-b9c1-2e8a47cb3f12" width="300px"/>|

But it might not be a correct fix, so please have a look at it 😊

mattleibow commented:
Hmmm, this looks like the #22520 issue so you fixed that :)

There is also #22521 which may also be related.

mattleibow commented:
This is nice that it fixes Android, I think iOS also has the same issue. How easy is it to fix ios and then we merge? If things are complicated, as buttons on iOS are, then we can do a separate PR.

I know TJ has been working on normal buttons and that has been a pain - especially since we can't derive from UIButton.

But no need to delay this fix for Android just because ios is being difficult. But, if there is basic padding math then it may be worth just doing it all in one go.

kubaflo commented:
> This is nice that it fixes Android, I think iOS also has the same issue. How easy is it to fix ios and then we merge? If things are complicated, as buttons on iOS are, then we can do a separate PR.
> 
> I know TJ has been working on normal buttons and that has been a pain - especially since we can't derive from UIButton.
> 
> But no need to delay this fix for Android just because ios is being difficult. But, if there is basic padding math then it may be worth just doing it all in one go.



> This is nice that it fixes Android, I think iOS also has the same issue. How easy is it to fix ios and then we merge? If things are complicated, as buttons on iOS are, then we can do a separate PR.
> 
> I know TJ has been working on normal buttons and that has been a pain - especially since we can't derive from UIButton.
> 
> But no need to delay this fix for Android just because ios is being difficult. But, if there is basic padding math then it may be worth just doing it all in one go.

@mattleibow Yeah, I've added a commit that fixes iOS

<img src="https://github.com/dotnet/maui/assets/42434498/1884910d-1b2a-4d2f-91ed-9b190a96cd70" width="300px"/>

mattleibow commented:
/azp run

mattleibow commented:
I am just looking at the results and the ones with small radii are looking nice, but the one with the big corners has a much smaller image. This is probably the more correct, but I am wondering if this is a breaking change or even what we want.

If I have say a 100x100 image, and a button with 0 border, 0 padding and 0 corner, then the button will be maybe 100x100 (ignoring any weird default paddings)

If I add a 10px border, then I would sort of be OK with the bounds increasing to 120x120. Similarly, if i add 10px padding, the 120x120 is expected.

But, if I add a massive corner, then all of a sudden my image shrinks. 

| 0 pad, border, corner | 25% padding | 25% border | 25% corner |
|:-:|:-:|:-:|:-:|
| ![](https://github.com/dotnet/maui/assets/1096616/66149c56-ab2c-482e-b226-8edb45f494c5) | ![](https://github.com/dotnet/maui/assets/1096616/4d75dfab-b276-41d7-8c56-767d0c000778) | ![](https://github.com/dotnet/maui/assets/1096616/dac93541-6fc4-4e1f-b4f6-5b7dabb14c6f) | ![](https://github.com/dotnet/maui/assets/1096616/15a0b94e-ce5a-4542-881b-a68b57c894c0) Current |
| | | | ![](https://github.com/dotnet/maui/assets/1096616/c336f3b4-7976-4397-963c-c895a3a3b102) Maybe? |

If we look at it with more corner:

| Base | 50% corner PR | 50% corner Maybe? |
|:-:|:-:|:-:|
| ![](https://github.com/dotnet/maui/assets/1096616/66149c56-ab2c-482e-b226-8edb45f494c5) | ![](https://github.com/dotnet/maui/assets/1096616/e50c33a4-0fb8-48b3-80e6-a3833cacc534) | ![](https://github.com/dotnet/maui/assets/1096616/615ec2e8-b3ee-4d68-b11f-a3cebf94030a) |

That sort of looks bad because of the image, but if we use a round image:

TODO

kubaflo commented:
Hmmm setting an additional padding based only on the border width seems to work on iOS 
`int additionalPadding = (int)platformButton.Layer.BorderWidth;`
,but when doing the same on Android, the images don't look good when the corner radius is big enough

|iOS|Android|
|--|--|
|<img src="https://github.com/dotnet/maui/assets/42434498/435598b9-3de7-47bf-943e-5ee4f869b273" width="300px"/>|<img src="https://github.com/dotnet/maui/assets/42434498/79184dc2-0c4e-446f-aaad-18f22e62ba2f" width="300px"/>|

kubaflo commented:
> Hmmm setting an additional padding based only on the border width seems to work on iOS `int additionalPadding = (int)platformButton.Layer.BorderWidth;` ,but when doing the same on Android, the images don't look good when the corner radius is big enough
> 
> iOS	Android
> <img alt="" width="300px" src="https://private-user-images.githubusercontent.com/42434498/335625963-435598b9-3de7-47bf-943e-5ee4f869b273.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTcxNjg2NjEsIm5iZiI6MTcxNzE2ODM2MSwicGF0aCI6Ii80MjQzNDQ5OC8zMzU2MjU5NjMtNDM1NTk4YjktM2RlNy00N2JmLTk0M2UtNWVlNGY4NjliMjczLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTMxVDE1MTI0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY4YmM1MmY2N2JhMzkzNjc1M2JiMzJiNWJjYWJkNDgzNmU2MGZjNjI1M2FmNTEwMzE5MjBiYjk1MmQ0YzEwNWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.GycDMjj3CkvRRosQZZpjekfBueCFumC-adIZDGgwdIc">	<img alt="" width="300px" src="https://private-user-images.githubusercontent.com/42434498/335627647-79184dc2-0c4e-446f-aaad-18f22e62ba2f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTcxNjg2NjEsIm5iZiI6MTcxNzE2ODM2MSwicGF0aCI6Ii80MjQzNDQ5OC8zMzU2Mjc2NDctNzkxODRkYzItMGM0ZS00NDZmLWFhYWQtMThmMjJlNjJiYTJmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTMxVDE1MTI0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdkNTE2MDg4Y2M0MWUyYzkwMjk4ZmU0OWFlM2YwYmQ0YzU4ODgyMGMwNDJlMTc0ZmYwMTYzZTY3ZjA0NmQ1MmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.-uKwxG251-79NuHgKsLyvd7_6iecRmd7c7pk74EPX4k">

@mattleibow I think Android would work if the border could overlay the image

mattleibow commented:
@kubaflo  I finished my comment lol - had to save it so I wouldn't lose it.

But yeah, the image looks weird with the shite, but mostly the image looks OK if the background is transparent.

But, this is maybe not so much the issue of what it looks like, but rather the predictability and intuitiveness.

For example:
* How would you make the ball image in my previous comment where the ball image is the same size as the button?
* When you get a big corner, how can you know what the "added padding" is if you don't know the algorithm.

When the button looks wrong, you know to add padding to make the image smaller. But, if we add the padding, devs have no way to remove when they actually want to clip.

The dotnet_bog image is bad because it is a solid color image, but most images will be some form of transparency where the corners do not matter.

Windows maybe looks OK because the border clips the contents?

We _could_ maybe have a look at wrapping the image in a clipping drawable. I think there is a MaskDrawable or something that we are using already to clip the ripple animation.

mattleibow commented:
I was playing with this:

```xml
<?xml version="1.0" encoding="utf-8" ?>
<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             x:Class="Maui.Controls.Sample.Issues.Issue20801"
             Title="Issue20801">
        <VerticalStackLayout
            Spacing="25"
            Padding="25">

            <ImageButton
                Source="beach_ball.png"
                BorderColor="Blue"
                BackgroundColor="Silver"
                HorizontalOptions="Center"
                HeightRequest="200"
                WidthRequest="200"
                Aspect="Fill"
                Padding="{Binding Value, Source={Reference PaddingSlider}, Converter={StaticResource ThicknessConverter}}"
                BorderWidth="{Binding Value, Source={Reference BorderWidthSlider}}"
                CornerRadius="{Binding Value, Source={Reference CornerRadiusSlider}}"
                AutomationId="imageButton"/>

            <Label Text="Padding" />
            <Slider
                x:Name="PaddingSlider"
                Maximum="200"
                Minimum="0"
                Value="0"/>

            <Label Text="Border Width" />
            <Slider
                x:Name="BorderWidthSlider"
                Maximum="200"
                Minimum="0"
                Value="0"/>

            <Label Text="Corner Radius" />
            <Slider
                x:Name="CornerRadiusSlider"
                Maximum="200"
                Minimum="0"
                Value="0"/>

        </VerticalStackLayout>
</ContentPage>
```

![beach_ball](https://github.com/dotnet/maui/assets/1096616/8bb7b2a7-a380-4868-aa0b-5d643ef79af8)

mattleibow commented:
I also notice that changing the padding updates the things correctly, but if I just change the width/corner then the padding calculations do not kick in.

When the border/corner changes, we may have to fire `Handler.UpdateValue(nameof(Padding))` in the mapper.

kubaflo commented:
> @kubaflo I finished my comment lol - had to save it so I wouldn't lose it.
> 
> But yeah, the image looks weird with the shite, but mostly the image looks OK if the background is transparent.
> 
> But, this is maybe not so much the issue of what it looks like, but rather the predictability and intuitiveness.
> 
> For example:
> 
> * How would you make the ball image in my previous comment where the ball image is the same size as the button?
> * When you get a big corner, how can you know what the "added padding" is if you don't know the algorithm.
> 
> When the button looks wrong, you know to add padding to make the image smaller. But, if we add the padding, devs have no way to remove when they actually want to clip.
> 
> The dotnet_bog image is bad because it is a solid color image, but most images will be some form of transparency where the corners do not matter.
> 
> Windows maybe looks OK because the border clips the contents?
> 
> We _could_ maybe have a look at wrapping the image in a clipping drawable. I think there is a MaskDrawable or something that we are using already to clip the ripple animation.

You're right! When the image inside the imageButton fills the available space, then devs can easily adjust the padding property to their needs. I've pushed a commit. I also added a commit that should solve the padding update

kubaflo commented:
@mattleibow so is there a need to look into it?

`We could maybe have a look at wrapping the image in a clipping drawable. I think there is a MaskDrawable or something that we are using already to clip the ripple animation.`

jsuarezruiz commented:
/azp run

jsuarezruiz commented:
/azp run

PureWeen commented:
/rebase

PureWeen commented:
/azp run

PureWeen commented:
/azp run

PureWeen commented:
/azp run

PureWeen commented:
/azp run

PureWeen commented:
/azp run

kubaflo committed:
ImageButton border (BorderWidth) overlaps the image

kubaflo committed:
Modified snapshots

PureWeen commented:
/azp run

jsuarezruiz committed:
More samples

jsuarezruiz committed:
Added another UITest

jsuarezruiz commented:
/azp run

jsuarezruiz committed:
Added pending snapshot

jsuarezruiz commented:
/azp run

jsuarezruiz commented:
/azp run

PureWeen commented:
- failing iOS controls tests not related
